### PR TITLE
Support more syntax

### PR DESCRIPTION
**Pull request by pplam**
### New supportings

- CTEs syntax;
- ALTER TABLE RENAME COLUMN syntax;
- LIMIT clause;
- DELETE statement;
- AT TIME ZONE expression;
- More json_access expressions;
- Support Dotted table names when creating index;
- Multi-words types (VARYING, PRECISION, WITH/WITHOUT TIME ZONE);
- Common mathematical operators;
- Unary operators;
- Conditional expressions;

    
### Improvements
- The ORDER BY clause syntax;
- The JOIN clause;
- The select subexpression;
- The UPDATE statement;
- The dotted_name & _identifier;
- Array type;
- Binary operators;
- Reorganize precedences;
- Refactor SELECT, UPDATE, INSERT, DELETE statements;

### CODE REVIEW

**Comment by m-novikov** [comment]
Thank you for the PR :)
This needs more tests as otherwise there is a probability of unexpected breakage of introduced syntax.
Existing tests can be updated with a command
```
tree-sitter test -u
```
It will update existing tests to match current output but use it sparingly.

File: grammar.js [hunk 1]
```diff
@@ -67,6 +84,32 @@ module.exports = grammar({
...
+      choice(
+        $.select_statement,
+        $.delete_statement,
+        $.insert_statement,
+        $.update_statement,
+      ),
+      ')',
+    ),
+
+    select_statement: $ => seq(optional($._ctes), $._select_statement),
```
Comment (m-novikov) [inline grammar.js:108]: I think that instead statements having an optional `WITH` clause, they should instead be included to CTE:
```js
cte: $ => seq(
      $.identifier,
      kw('AS'),
      optional(seq(optional(kw('NOT')), kw('MATERIALIZED'))),
      '(',
      choice(
        $.select_statement,
        $.delete_statement,
        $.insert_statement,
        $.update_statement,
      ),
      ')',
      // optional?
      choice(
        $.select_statement,
        $.delete_statement,
        $.insert_statement,
        $.update_statement,
      ),
    ),
```

Comment (pplam) [reply grammar.js:108]: According to the docs, in fact, the `SELECT`, `UPDATE`, `DELETE` and `INSERT` statements may have an optional `WITH` clause.

And the `WITH` clause may contains multiple `with_query`s (the `cte` node here).

Ref: https://www.postgresql.org/docs/14/sql-insert.html

**Comment by m-novikov** [comment]
Hey, I took a couple of patches from your PR attributed to you in the commits e.g https://github.com/m-novikov/tree-sitter-sql/commit/0d3cde974e5376fc89e87bf34905b1c4f242d8fe Thanks a lot.

Currently, I am working on an external parser for dollar-quoted strings.

Overall it would be simpler for me to merge stuff with smaller PRs, because I also working on some parts and for example unary expressions were helpful.

**Comment by pplam** [comment]
It's good practice to create small PRs, next time i will submit as small PR as possible

**Comment by pplam** [comment]
> Thank you for the PR :) This needs more tests as otherwise there is a probability of unexpected breakage of introduced syntax. Existing tests can be updated with a command
> 
> ```
> tree-sitter test -u
> ```
> 
> It will update existing tests to match current output but use it sparingly.

Hi @m-novikov , I just resolved all conflicts and rewrote the git history. 

And also add test for each changed/added rule.

Please review this PR again, thanks!

**Pull request merged**